dafny/lcf/datalog/AstBuilder.cs

```diff
@@ -70,16 +70,88 @@ internal class AstBuilder : datalogBaseVisitor<object>
+        case "member":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Member(), terms);
+        case "lists:reverse":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Reverse(), terms);
+        case "reverse":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Reverse(), terms);
+        case "lists:nth1":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Nth1(), terms);
+        case "nth1":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Nth1(), terms);
+        default:
+          return null;
+      }
+    }
+
+    public override object VisitExpression(datalogParser.ExpressionContext context) {
+      var left = (_module.Term)VisitTerm(context.left);
+      var right = (_module.Term)VisitTerm(context.right);
+      var terms = new List<_module.Term> { left, right };
+      var terms2 = (Dafny.ISequence<_ITerm>) Sequence<_module.Term>.Create(terms.Count, i => terms[(int) i]);
+      switch(context.name.Text) {
+        case "=<":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatLeq(), terms2);
+        case ">=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatGeq(), terms2);
+        case "\\=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatNeq(), terms2);
+        case "=\\=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatNeq(), terms2);
+        case "<":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatLt(), terms2);
+        case ">":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatGt(), terms2);
+        default:
+          return new _module.Prop_Eq(left, right);
+      }
+    }
+
+    public override object VisitApp(datalogParser.AppContext context) {
+      var name = Sequence<char>.FromString(context.name.Text);
+      if (context.term_list() == null ) {
+        var terms = new List<_module.Term>();
+        var terms2 = Sequence<_module.Term>.Create(terms.Count, i => terms[(int) i]);
+        return new _module.Prop_App(name, (Dafny.ISequence<_ITerm>) terms2);
+      } else {
+        var terms = (Dafny.ISequence<_ITerm>) VisitTerm_list(context.term_list());
+        return new _module.Prop_App(name, terms);
+      }
     }
 
     public override object VisitAtom(datalogParser.AtomContext context) {
       return new Term_Const(new Const_Atom(Sequence<char>.FromString(context.val.Text)));
     }
 
+    public override object VisitNatural(datalogParser.NaturalContext context) {
+      return new Term_Const(new Const_Nat(BigInteger.Parse(context.numeral.Text)));
+    }
+
+    public override object VisitString(datalogParser.StringContext context) {
+      return new Term_Const(new Const_Str(Sequence<char>.FromString(context.s.Text.Substring(1, context.s.Text.Length - 2))));
+      // A substring of the text is taken in order to remove the double quotes.
```

Comment:
Dod you say you had supported escaping now?

Reply:
I fixed it to replace escaped characters with the actual characters. However, I discovered that adding support for escaped double quotes broke the parser, so I had to stop supporting them for now (although other escaped characters are now fully supported).